Handle cases in EventTable.cluster() where table is empty or clustered multiple times

myNameIsPatrick opened the pull request:
This PR handles a few edge cases more gracefully when calling `EventTable.cluster()`:

1. The table is empty.
2. The table is already clustered.

In both cases, the error encountered looks like:

```
>>> table.cluster('time', 'snr', 0.1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/patrick.godwin/workspace/projects/gwpy/gwpy/table/table.py", line 782, in cluster
    padded_sublists = [numpy.append(s, numpy.array([s[-1]+1]))
  File "/home/patrick.godwin/workspace/projects/gwpy/gwpy/table/table.py", line 782, in <listcomp>
    padded_sublists = [numpy.append(s, numpy.array([s[-1]+1]))
IndexError: index -1 is out of bounds for axis 0 with size 0
```

In both cases, the original table is now returned as a result. I have also added a test which checks that running `table.cluster()` with the same parameters twice returns the same table.

myNameIsPatrick commented:
After looking this over again, I realize I am returning the same table rather than a new table in both cases. I should easily be able to modify it depending on what the desired behavior is.

duncanmmacleod commented on gwpy/table/table.py:
```diff
@@ -779,14 +779,23 @@ def cluster(self, index, rank, window):
 
+        # If no rows, no need to cluster
+        if len(self) == 0:
+            return self
```
I like this short-circuit, however the documented behvaiour is to return a _new_ table, so we might need to make a copy:

```suggestion
            return self.copy()
```

duncanmmacleod commented on gwpy/table/table.py:
```diff
@@ -779,14 +779,23 @@ def cluster(self, index, rank, window):
+
+        # If no such cluster points, no need to cluster
+        if len(clusterpoints) == 0:
+            return self
```
As above, maybe this:

```suggestion
            return self.copy()
```

duncanmmacleod reviewed (changes requested):
I think this is great, sorry it has taken a ludicrous amount of time to review, just one pedantic change.

myNameIsPatrick commented:
Thanks @duncanmmacleod and no worries. I've changed it so it returns copies in both cases now. I also noticed that one of the newly introduced lines wasn't covered by the test suite, so I added an additional test to handle the empty table case.
